### PR TITLE
Add fix qtpie/reaxff

### LINKED ISSUE #4226: [Feature Request] Allow for QTPIE charge equilibration with external efield
**Summary**

LAMMPS currently supports `fix qeq/reaxff` and `fix acks2/reaxff`, which allow for charge equilibration with an external electric field. It could be valuable to add `fix qtpie/reaxff` to the REAXFF package to allow for charge equilibration using the [QTPIE](https://doi.org/10.1016/j.cplett.2007.02.065) charge equilibration scheme. 

**Detailed Description**

When using [QEq](https://pubs.acs.org/doi/10.1021/j100161a070) charge equilibration with pair_style reaxff and when applying an external electric field, the dynamics of the simulation become unrealistic due to long range charge transfers predicted by the QEq method. For example, molecules that should be neutral acquire a net charge and are driven to translate by the electric field. This has been discussed by [Kritikos et al](https://doi.org/10.1016/j.combustflame.2022.112238) and [Koski et al](https://pubs.acs.org/doi/10.1021/acs.jctc.1c00975). QTPIE charge equilibration can successfully shield charge transfers (see, e.g., [Kritikos et al](https://pubs.acs.org/doi/full/10.1021/acs.jpca.0c08040)) and has a similar formulation to QEq, which means that it can be, to a reasonable approximation, used with the same REAXFF that QEq has been parameterised for.  Furthermore, QTPIE can be formulated to allow for periodic boundaries in the direction of an external electric field.

I would like to try and implement this myself. I saw that work has been completed on this before by @stanmoore1 at https://github.com/stanmoore1/lammps/tree/qtpie . I wanted to know if there were any particular obstacles that prevented the implementation of QTPIE from being completed or anything in particular that I should be aware of before giving it a go?

Thanks!

### PR DESCRIPTION
**Pull request by navlalli**
**Summary**

`fix qtpie/reaxff` performs charge equilibration using the [QTPIE](https://doi.org/10.1016/j.cplett.2007.02.065) charge equilibration scheme when using `pair_style reaxff`. This fix can successfully shield long range charge transfer commonly seen with the QEq charge equilibration scheme and allows for periodic boundaries in the direction of an external electric field.

**Related Issue(s)**

Closes #4226 

**Author(s)**

Navraj S Lalli (Imperial College London)
Efstratios M Kritikos (California Institute of Technology)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Changes in this pull request will not break backward compatibility.

**Implementation Notes**

This fix is implemented by adding fix_qtpie_reaxff.cpp and fix_qtpie_reaxff.h to the REAXFF package.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

### CODE REVIEW

**Comment by akohlmey** [comment]
@navlalli please make sure you click on "Ready for review" when you want people to look at your pull request or put a comment here, possibly with an @ mention of the developer that you want to consult.

**Comment by akohlmey** [comment]
@navlalli FYI, your code has whitespace issues (see the details of the `dev/pull_requests/checkstyle` test). To resolve those, please go to the `src` folder, type `make fix-whitespace` and commit the changes.

**Comment by akohlmey** [comment]
@navlalli FYI, I just merged your pull request branch with upstream because we are adding two more mandatory tests that need to be passed by all contributions. So before you make any more changes on your side, be certain to pull before your push. :wink:

**Comment by stanmoore1** [comment]
@navlalli is this PR ready for review or are you still adding more commits?

**Comment by navlalli** [comment]
@stanmoore1 Ready for review!

File: doc/src/fix_qtpie_reaxff.rst [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+Restrictions
+""""""""""""
+
+This fix is part of the REAXFF package.  It is only enabled if
+LAMMPS was built with that package. See the :doc:`Build package
+<Build_package>` page for more info.
+
+This fix does not correctly handle interactions involving multiple
+periodic images of the same atom.  Hence, it should not be used for
+periodic cell dimensions less than 10 Angstroms.
```
Comment (stanmoore1) [inline doc/src/fix_qtpie_reaxff.rst]: Wouldn't this depend on the cutoff? Could say something like "Hence, it should not be used for
periodic cell dimensions less than the pairwise cutoff, typically 10 Angstroms."

File: doc/src/fix_qtpie_reaxff.rst [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+Equal-style variables can be used for electric field vector
+components without any further settings. Atom-style variables can be used
+for spatially-varying electric field vector components, but the resulting
+electric potential must be specified as an atom-style variable using
+the *potential* keyword for `fix efield`.
+
+Related commands
+""""""""""""""""
+
+:doc:`pair_style reaxff <pair_reaxff>`, :doc:`fix qeq/reaxff <fix_qeq_reaxff>`
```
Comment (stanmoore1) [inline doc/src/fix_qtpie_reaxff.rst]: Need to list `fix acks2/reaxff` as a related command too

File: src/REAXFF/fix_qeq_reaxff.cpp [hunk 3]
```diff
@@ -1158,7 +1158,7 @@ void FixQEqReaxFF::get_chi_field()
     for (int i = 0; i < nlocal; i++) {
       if (mask[i] & efgroupbit) {
         if (region && !region->match(x[i][0],x[i][1],x[i][2])) continue;
-        chi_field[i] = -efield->efield[i][3];
+        chi_field[i] = efield->efield[i][3];
```
Comment (stanmoore1) [inline src/REAXFF/fix_qeq_reaxff.cpp:1161]: Does this need to be propagated to other styles like ACKS2 and KOKKOS?

**Review by stanmoore1 — CHANGES_REQUESTED** [review]
@navlalli overall this looks really good, I had a few comments/questions

File: doc/src/fix_qtpie_reaxff.rst (hunk 1 above)
Comment (navlalli) [reply doc/src/fix_qtpie_reaxff.rst]: Good point. I have updated the documentation (I have also added this change to the documentation of fix qeq/reaxff and fix acks2/reaxff).

File: src/REAXFF/fix_qeq_reaxff.cpp (hunk 3 above)
Comment (navlalli) [reply src/REAXFF/fix_qeq_reaxff.cpp:1161]: It seems that acks2/reaxff and the KOKKOS versions of qeq/reaxff and acks2/reaxff call `get_chi_field()` from `FixQEqReaxFF`, so I don't think any further changes are needed.

**Comment by navlalli** [comment]
@stanmoore1 Thank you for the review! I believe I have addressed the points. Please let me know if there is anything else.

File: src/REAXFF/fix_qtpie_reaxff.cpp [hunk 4]
```diff
@@ -0,0 +1,1231 @@
...
+
+  if (pack_flag == 1)
+    for (m = 0; m < n; m++) buf[m] = d[list[m]];
+  else if (pack_flag == 2)
+    for (m = 0; m < n; m++) buf[m] = s[list[m]];
+  else if (pack_flag == 3)
+    for (m = 0; m < n; m++) buf[m] = t[list[m]];
+  else if (pack_flag == 4)
+    for (m = 0; m < n; m++) buf[m] = atom->q[list[m]];
+  else if (pack_flag == 5) {
```
Comment (jtclemm) [inline src/REAXFF/fix_qtpie_reaxff.cpp]: Is this option ever used? If so, the *comm_forward = 1* assignment in the constructor is insufficient. Ditto for reverse comm.

File: src/REAXFF/fix_qtpie_reaxff.cpp [hunk 5]
```diff
@@ -0,0 +1,1231 @@
...
+  // in calc_chi_eff()
+  const double exp_min = find_min_exp(gauss_exp,ntypes+1);
+  const int olap_cut = 10; // overlap integrals are neglected if less than pow(10,-olap_cut)
+  dist_cutoff = sqrt(2*olap_cut/exp_min*log(10.0));
+
+  // read chi, eta and gamma
+
+  if (utils::strmatch(arg,"^reaxff")) {
+    reaxflag = 1;
+    Pair *pair = force->pair_match("^reaxff",0);
```
Comment (jtclemm) [inline src/REAXFF/fix_qtpie_reaxff.cpp:242]: I understand this is called in the post constructor. If so, does this fix's behavior change if I defined a reaxff pair style after this fix is defined in the input script?

Comment (navlalli) [reply src/REAXFF/fix_qtpie_reaxff.cpp:242]: Correct. One gets `ERROR: No reaxff pair style for fix qtpie/reaxff (src/REAXFF/fix_qtpie_reaxff.cpp:243)` if you put `fix qtpie/reaxff` before defining the reaxff pair style. This behaviour is in alignment with `fix qeq/reaxff` and therefore `fix acks2/reaxff`.

File: src/REAXFF/fix_qtpie_reaxff.cpp (hunk 4 above)
Comment (navlalli) [reply src/REAXFF/fix_qtpie_reaxff.cpp]: This is a good point. The `pack_flag = 5` option is never used (it is only used in `fix qeq/reaxff/omp`) and has now been removed.

**Review by stanmoore1 — APPROVED** [review]
I approve

**Review by jtclemm — APPROVED** [review]

**Pull request merged**